{{cookiecutter.module_name}}/main.py

```diff
@@ -1,94 +1,74 @@
-        print ( "IoT Hub Client for Python" )
-
-        hub_manager = HubManager(protocol)
-
-        print ( "Starting the IoT Hub Python sample using protocol %s..." % hub_manager.client_protocol )
-        print ( "The sample is now waiting for messages and will indefinitely.  Press Ctrl-C to exit. ")
-
-        while True:
-            time.sleep(1)
-
-    except IoTHubError as iothub_error:
-        print ( "Unexpected error %s from IoTHub" % iothub_error )
-        return
-    except KeyboardInterrupt:
-        print ( "IoTHubModuleClient sample stopped" )
-
-if __name__ == '__main__':
-    main(PROTOCOL)
\ No newline at end of file
+# Copyright (c) Microsoft. All rights reserved.
+# Licensed under the MIT license. See LICENSE file in the project root for
+# full license information.
+
+import time
+import os
+import sys
+import asyncio
+from six.moves import input
+import threading
+from azure.iot.device.aio import IoTHubModuleClient
+
+async def main():
+    try:
+        print ( "\nPython %s\n" % sys.version )
+        print ( "IoT Hub Client for Python" )
+
+        print ( "The sample is now waiting for messages and will indefinitely. ")
+
+        # The client object is used to interact with your Azure IoT hub.
+        module_client = IoTHubModuleClient.create_from_edge_environment()
+
+        # connect the client.
+        await module_client.connect()
+
+        # define behavior for receiving an input message on input1
+        async def input1_listener(module_client):
+            while True:
+                input_message = await module_client.receive_message_on_input("input1")  # blocking call
+                print("the data in the message received on input1 was ")
+                print(input_message.data)
+                print("custom properties are")
+                print(input_message.custom_properties)
+                print("forwarding mesage to output1")
+                await module_client.send_message_to_output(input_message, "output1")
+
+        # define behavior for halting the application
+        def stdin_listener():
+            while True:
+                try:
+                    selection = input("Press Q to quit\n")
+                    if selection == "Q" or selection == "q":
+                        print("Quitting...")
+                        break
+                except:
+                    time.sleep(10)
+
+        # Schedule task for C2D Listener
+        listeners = asyncio.gather(input1_listener(module_client))
+
+        # Run the stdin listener in the event loop
+        loop = asyncio.get_event_loop()
+        user_finished = loop.run_in_executor(None, stdin_listener)
+
+        # Wait for user to indicate they are done listening for messages
+        await user_finished
+
+        # Cancel listening
+        listeners.cancel()
+
+        # Finally, disconnect
+        await module_client.disconnect()
+
+    except Exception as e:
+        print ( "Unexpected error %s " % e )

```

Comment:
if exception, we didn't throw out the exception and exit?

Reply:
I'm following existing python template behavior. The module will exit after printed the error message. Do you mean we throw an exception here?

Reply:
If exit normally, that means if the outside restart policy is on-failure, it will not work